### PR TITLE
Use new snyk protect

### PR DESCRIPTION
**Pull request by maxjeffos**
Use the new `@snyk/protect` package to apply Snyk patches rather than the entire `snyk` package.

### CODE REVIEW

**Comment by aviadhahami** [comment]
@maxjeffos should this still be under @snyk/boost ?

**Review by jahed-snyk — APPROVED** [review]

File: lib/index.js [hunk 1]
```diff
@@ -49,12 +49,13 @@ function addProtect(pkg, cmdScript, packageManager = 'npm') {
...
   const protecting = existingScript && (pkg.scripts[existingScript].indexOf(protectCmd) !== -1);
 
-  // don't add anything if there is already protect command
   if (!protecting) {
-    pkg.scripts['snyk-protect'] = 'snyk protect';
     pkg.scripts[cmdScript] = getNewScriptContent(pkg.scripts[cmdScript], protectCmd);
   }
 
+  // update to the new snyk-protect script
+  pkg.scripts['snyk-protect'] = 'npm_config_yes=true npx @snyk/protect@latest"';
```
Comment (maxjeffos) [inline lib/index.js:57]: Setting the env var `npm_config_yes=true` rather than using `--yes` to be back-compatible with npm@6.

Comment (jahed-snyk) [reply lib/index.js:57]: Is that needed? I thought npm@6 doesn't prompt for confirmation? 🤔

Comment (maxjeffos) [reply lib/index.js:57]: It doesn't, but with npm@6, if you pass the `--yes` it doesn't work properly:
```
❯ npx --yes snyk@1.605.0 --version
6.14.12
```
That's the npm version, not the snyk version!

So in order to make it suppress the prompt on npm@7 and not cause this weirdness on npm@6, we can set the env var `npm_config_yes` - it is simply ignored by npm@6.

**Comment by maxjeffos** [comment]
> @maxjeffos should this still be under @snyk/boost ?

@orsagie moved it to loki.

**Review by aviadhahami — APPROVED** [review]

**Review by jahed-snyk — APPROVED** [review]

File: lib/index.js [hunk 2]
```diff
@@ -49,12 +49,13 @@ function addProtect(pkg, cmdScript, packageManager = 'npm') {
...
   const protecting = existingScript && (pkg.scripts[existingScript].indexOf(protectCmd) !== -1);
 
-  // don't add anything if there is already protect command
   if (!protecting) {
-    pkg.scripts['snyk-protect'] = 'snyk protect';
     pkg.scripts[cmdScript] = getNewScriptContent(pkg.scripts[cmdScript], protectCmd);
   }
 
+  // update to the new snyk-protect script
+  pkg.scripts['snyk-protect'] = 'npm_config_yes=true npx @snyk/protect@latest';
```
Comment (jahed-snyk) [inline lib/index.js:57]: env prefixes don't work on Windows.

File: test/npm.test.js [hunk 3]
```diff
@@ -23,24 +24,29 @@ test('add(test)', t => {
...
 test('add(protect)', t => {
   const pkg = getPkg();
-
   lib.add(pkg, 'protect', v);
   t.match(pkg.scripts.prepare, 'npm run snyk-protect', 'contains protect command');
   t.ok(!pkg.scripts.prepublish, 'does not contain prepublish');
-
-  t.equal(pkg.dependencies.snyk, '^' + v, 'includes snyk and latest');
+  t.equal(pkg.scripts['snyk-protect'], 'snyk-protect', 'snyk-protect is defined as `snyk-protect`');
+  testUtils.assertPackageNotDefined(t, pkg, 'snyk', 'dependencies');
```
Comment (jahed-snyk) [inline test/npm.test.js:31]: It's difficult to reason about these tests as they're asserting on too many things. Should we move the dependency assertions to their own tests (possibly unit tests for `updateSnykDependency`)?
If we want higher-level input/output package.json tests we can use snapshot testing.

File: test/npm.test.js [hunk 4]
```diff
@@ -132,17 +155,35 @@ test('add(test && protect) on empty package', t => {
   t.end();
 });
 
-
-test('already testing moves to prod deps when protect', t => {
+test('already testing moves stays with devDependencies when adding protect but upgrades version', t => {
```
Comment (jahed-snyk) [inline test/npm.test.js:158]: This sentence doesn't make sense. "moves stays"

File: lib/index.js [hunk 5]
```diff
@@ -122,37 +121,58 @@ function updateSnykVersion(pkg, version, packageManager = 'npm') {
...
     key = devDependencies;
   }
 
-  // if protecting, then always move snyk out of devdeps and optional
-  if (protecting) {
-    pkg.snyk = true;
-    if (key === devDependencies || key === optionalDependencies) {
-      delete pkg[key].snyk;
-      key = null;
+  if (testing) {
```
Comment (jahed-snyk) [inline lib/index.js:132]: Should we move all of the code above that calculates the value of `key` into this `if (testing)` statement? It's all useless outside the scope of this condition.

File: test/unit-tests.jest.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
+const lib = require("../lib/index");
```
Comment (jahed-snyk) [inline test/unit-tests.jest.js:1]: Can we have a test module for each function under test rather than a single large one?

File: test/yarn.test.js [hunk 7]
```diff
@@ -1,5 +1,6 @@
 const { test } = require('tap');
```
Comment (jahed-snyk) [inline test/yarn.test.js:1]: Instead of repeating the entire test suite for every packageManager, should we remove `yarn.test.js` and just have a single test under `npm.test.js` (rename that to be more generic) that ensures we are taking `packageManager` and applying it appropriately?

File: lib/index.js [hunk 8]
```diff
@@ -7,7 +7,9 @@ const optionalDependencies = 'optionalDependencies';
 
 module.exports = {
   add,
-  updateSnykVersion,
+  updateSnykVersion: updateSnykDependency,
```
Comment (jahed-snyk) [inline lib/index.js:10]: Why do we rename `updateSnykDependency` to `updateSnykVersion`?

If it's for backwards compatibility, it doesn't look like anyone's using this function outside of this module.

https://github.com/search?q=org%3Asnyk+updateSnykVersion&type=code

Also, since we're releasing a breaking change, it doesn't matter if we change the public interface that much (as long as we document it).

File: jest.config.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
+module.exports = {
```
Comment (jahed-snyk) [inline jest.config.js:1]: `testEnvironment`, `collectCoverage` (and related config keys) are defaults so we can remove them.

https://jestjs.io/docs/configuration#testenvironment-string
https://jestjs.io/docs/configuration#collectcoverage-boolean

File: lib/index.js [hunk 10]
```diff
@@ -49,12 +51,13 @@ function addProtect(pkg, cmdScript, packageManager = 'npm') {
   }
   const protecting = existingScript && (pkg.scripts[existingScript].indexOf(protectCmd) !== -1);
 
-  // don't add anything if there is already protect command
   if (!protecting) {
-    pkg.scripts['snyk-protect'] = 'snyk protect';
     pkg.scripts[cmdScript] = getNewScriptContent(pkg.scripts[cmdScript], protectCmd);
   }
 
+  // update to the new snyk-protect script
```
Comment (jahed-snyk) [inline lib/index.js:58]: Do we need this comment?

File: package.json [hunk 11]
```diff
@@ -13,12 +13,14 @@
...
     "cover": "tap -R spec --cov --coverage-report=lcov test/*.test.js",
     "test": "tap -R spec test/*.test.js",
+    "test:jest": "jest",
     "release": "semantic-release"
   },
   "keywords": [],
   "author": "snyk.io",
   "license": "Apache-2.0",
   "devDependencies": {
+    "jest": "^27.0.4",
```
Comment (jahed-snyk) [inline package.json:23]: How much work is it to migrate the tap tests to jest so that we don't have to deal with fragmentation?

**Review by jahed-snyk — CHANGES_REQUESTED** [review]

File: lib/index.js [hunk 12]
```diff
@@ -101,18 +105,13 @@ function add(pkg, type, version, cmdScript, packageManager) {
   }
 
   if (version) {
-    updateSnykVersion(pkg, version, packageManager);
+    const protecting = isProtecting(pkg, packageManager);
```
[the remark this answers is not in the document]
Comment (maxjeffos) [reply lib/index.js:107]: resolved

File: test/unit-tests.jest.js (hunk 6 above)
Comment (maxjeffos) [reply test/unit-tests.jest.js]: We could, but that wouldn't be in keeping with the test file per module thing.
Let me know if you like how I restructured the tests in terms of file layout.

File: test/yarn.test.js (hunk 7 above)
Comment (maxjeffos) [reply test/yarn.test.js]: I skipped this because it's a lot of yak shaving at this point and what we have works. If we're going to do that, I'd rather do it in a separate PR so that we aren't blocking releasing protect on it.

File: package.json (hunk 11 above)
Comment (maxjeffos) [reply package.json:24]: Yes! But I did it.

File: test/npm.test.js [hunk 13]
```diff
@@ -23,24 +24,29 @@ test('add(test)', t => {
 
 test('add(protect)', t => {
```
[the remark this answers is not in the document]
Comment (maxjeffos) [reply test/npm.test.js]: done

File: lib/index.js [hunk 14]
```diff
@@ -101,58 +104,70 @@ function add(pkg, type, version, cmdScript, packageManager) {
...
   const scripts = pkg.scripts || {};
-  const testing = (scripts.test || '').includes('snyk test');
   const protectCmd = packageManager + ' run snyk-protect';
-  const protecting = [
+  const protectingScript = [
     'prepare',
     'prepublish',
     'postinstall',
   ].find(key => (scripts[key] || '').includes(protectCmd));
+  return Boolean(protectingScript);
```
Comment (jahed-snyk) [inline lib/index.js:158]: Can use `.some` instead of `.find` to avoid the boolean conversion here.

File: test/unit/lib/index.test.js [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+      scripts: {
+        postinstall: "npm run snyk-protect",
+      },
+    };
+    const protecting = lib.isProtecting(pkg, "npm");
+    expect(protecting).toBe(true);
+  });
+});
+
+describe("updateSnykProtectDependency", () => {
```
Comment (jahed-snyk) [inline test/unit/lib/index.test.js:43]: My original suggestion also involved splitting these functions into their own modules to avoid a large unit test file with multiple responsibilities.

**Review by jahed-snyk — APPROVED** [review]
Left a comment but feel free to merge.

File: test/unit-tests.jest.js (hunk 6 above)
Comment (jahed-snyk) [reply test/unit-tests.jest.js]: I think it is, for readability and separation of concerns. But I'll leave the final decision up to you.

**Review by aviadhahami — APPROVED** [review]
